fsm/client.go

```diff
@@ -0,0 +1,154 @@
+package fsm
+
+import (
+	"fmt"
+	"log"
+
+	"time"
+
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/gen/swf"
+	"github.com/juju/errors"
+	. "github.com/sclasen/swfsm/sugar"
+)
+
+type FSMClient interface {
+	GetState(id string) (string, interface{}, error)
+	Signal(id string, signal string, input interface{}) error
+	Start(startTemplate swf.StartWorkflowExecutionInput, id string, input interface{}) (*swf.Run, error)
+}
+
+type ClientSWFOps interface {
+	ListOpenWorkflowExecutions(req *swf.ListOpenWorkflowExecutionsInput) (resp *swf.WorkflowExecutionInfos, err error)
+	ListClosedWorkflowExecutions(req *swf.ListClosedWorkflowExecutionsInput) (resp *swf.WorkflowExecutionInfos, err error)
+	GetWorkflowExecutionHistory(req *swf.GetWorkflowExecutionHistoryInput) (resp *swf.History, err error)
+	SignalWorkflowExecution(req *swf.SignalWorkflowExecutionInput) (err error)
+	StartWorkflowExecution(req *swf.StartWorkflowExecutionInput) (resp *swf.Run, err error)
+	TerminateWorkflowExecution(req *swf.TerminateWorkflowExecutionInput) (err error)
+	RequestCancelWorkflowExecution(req *swf.RequestCancelWorkflowExecutionInput) (err error)
+}
+
+func NewFSMClient(f *FSM, c ClientSWFOps) FSMClient {
+	return &client{
+		f: f,
+		c: c,
+	}
+}
+
+type client struct {
+	f *FSM
+	c ClientSWFOps
+}
+
+func (c *client) GetState(id string) (string, interface{}, error) {
+	var execution *swf.WorkflowExecution
+	open, err := c.c.ListOpenWorkflowExecutions(&swf.ListOpenWorkflowExecutionsInput{
+		Domain:          S(c.f.Domain),
+		MaximumPageSize: aws.Integer(1),
+		StartTimeFilter: &swf.ExecutionTimeFilter{OldestDate: &aws.UnixTimestamp{time.Unix(0, 0)}},
+		ExecutionFilter: &swf.WorkflowExecutionFilter{
+			WorkflowID: S(id),
+		},
+	})
+
+	if err != nil {
+		if ae, ok := err.(aws.APIError); ok {
+			log.Printf("component=client fn=GetState at=list-open error-type=%s message=%s", ae.Type, ae.Message)
+		} else {
+			log.Printf("component=client fn=GetState at=list-open error=%s", err)
+		}
+		return "", nil, err
+	}
+
+	if len(open.ExecutionInfos) > 0 {
+		execution = open.ExecutionInfos[0].Execution
+	} else {
+		closed, err := c.c.ListOpenWorkflowExecutions(&swf.ListOpenWorkflowExecutionsInput{
```

Comment:
Yep should be closed and no, it needs to be a ListClosedWorkflowExecutionsInput though the fields are identical :dizzy_face: